### PR TITLE
feat(search): Hnsw Vector Search Optimizaton Pass

### PR DESCRIPTION
**Pull request by Beihao-Zhou**
This PR is part of #2426 

## Summary

- Add optimization pass to transform `sort by` to `knn search expr` for SQL query
- Use cost model to prioritize KNN search and range search
- Add vector range search in `Filter` executor in case that vector range search acts as the prefilter for KNN search
- Support prefilters in redis query transformer

### CODE REVIEW

File: src/search/passes/transfer_to_knn.h
Comment (PragmaTwice) [inline src/search/passes/transfer_to_knn.h]: Seems here we cannot guarantee that the semantics of LIMIT clause is satisfied, e.g.
```sql
SELECT * FROM idx WHERE num > 1 ORDER BY vec <-> [...] LIMIT 10;
```
for this SQL statement, users want 10 records where `num > 1` (unless there's less than 10 records satisfying `num > 1`), 

but after transformation it's `(KNN(10) vec <-> [...]) AND num > 1`, which implies the number of records may be less than 10 even if there's more than 10 records satisfying `num > 1`, since it equals to filtering `num > 1` from `KNN(10)` query (firstly get 10 records from KNN search and then filter records of `num <= 1` out of these 10).

Comment (PragmaTwice) [reply src/search/passes/transfer_to_knn.h]: And in the RediSearch query part, I think it's same, i.e. we cannot simply treat prefilter KNN search as AND expressions since it requries a specific number of records (`k`) -- and it breaks the semantics of `AND`.

So if we cannot currently come up with a good solution, we can prohibit any KNN search with prefilter (in transformer or sema checker) and leave it to further PRs.

**Review by PragmaTwice — COMMENTED** [review]
The rest looks good to me!

Also if you have time maybe you can also add [some golang test cases](https://github.com/apache/kvrocks/blob/unstable/tests/gocase/unit/search/search_test.go) for vector search, after this PR is merged.

File: src/search/passes/transfer_to_knn.h
Comment (Beihao-Zhou) [reply src/search/passes/transfer_to_knn.h]: Thanks for pointing it out! Will prohibit hybrid query from this PR then. Also open to any suggestions from the community here <3

Comment (Beihao-Zhou) [reply src/search/passes/transfer_to_knn.h]: Ready for review again! Prohibit hybrid query in transformer for Redis and semantic check for SQL.

File: src/search/passes/transfer_to_knn.h
Comment (PragmaTwice) [inline src/search/passes/transfer_to_knn.h]:
```suggestion
struct SortByWithLimitToKnnExpr : Visitor {
```

File: src/search/passes/transfer_to_knn.h
Comment (PragmaTwice) [inline src/search/passes/transfer_to_knn.h]: How about `limit->offset` here? Seems it's ignored.

Comment (Beihao-Zhou) [reply src/search/passes/transfer_to_knn.h]: Oh thanks for your reminder! I'll wrap a Limit executor for all Knn scan then, passing `offset + count` as `k` to `KnnScan`.

Comment (Beihao-Zhou) [reply src/search/passes/transfer_to_knn.h]: Done <3

File: tests/cppunit/ir_pass_test.cc
Comment (PragmaTwice) [inline tests/cppunit/ir_pass_test.cc:120]: I'm wondering if this query is valid and has a correct execution result 🤔

Comment (Beihao-Zhou) [reply tests/cppunit/ir_pass_test.cc:120]: For `"select a from b where false order by embedding <-> [3,1,2] limit 5"`, this will be set as `noop` in `LowerToPlan`. 
Ref: https://github.com/apache/kvrocks/pull/2466#discussion_r1705921930

File: tests/cppunit/ir_pass_test.cc
Comment (Beihao-Zhou) [inline tests/cppunit/ir_pass_test.cc:135]: Referring to this test case

Comment (PragmaTwice) [reply tests/cppunit/ir_pass_test.cc:135]: Ahh sorry, got it now. Thank you!

**Review by PragmaTwice — APPROVED** [review]

**Pull request merged**
